tsl/src/continuous_aggs/invalidation_threshold.c

```diff
@@ -0,0 +1,272 @@
+ *
+ * Any update to a threshold (either via an update or insert) requires
+ * completely exclusive access (including blocking readers). This is necessary
+ * to ensure we do not lose invalidations; all transactions that read the
+ * invalidation threshold must have either completed, or not yet read the
+ * value. If a weaker lock was used, an insert transaction that is concurrent
+ * with a refresh could race to read the old threshold before it was updated
+ * and invalidations written during the insert (but after the update) would be
+ * lost.
+ *
+ * Since the invalidation threshold is currently stored in a shared table, all
+ * INSERTs (that read the threshold) and refreshes of continuous aggregates
+ * will be blocked. Therefore, we try to avoid heavy locking if possible by
+ * first reading the threshold with a read lock. If the threshold needs no
+ * update, we need not grab any further locks.
+ */
+bool
+continuous_agg_invalidation_threshold_set(int32 hyper_id, int64 new_threshold)
+{
+	Catalog *catalog = ts_catalog_get();
+	Oid threshold_relid = catalog_get_table_id(catalog, CONTINUOUS_AGGS_INVALIDATION_THRESHOLD);
+	Relation rel;
+	ItemPointerData tid;
+	int64 threshold;
+	bool threshold_written = false;
+
+	MemSet(&tid, 0, sizeof(tid));
+	Assert(!ItemPointerIsValid(&tid));
+	threshold = invalidation_threshold_read(hyper_id, &tid, AccessShareLock);
+
+	if (ItemPointerIsValid(&tid))
+	{
+		/* Found existing threshold. Check if update is needed, otherwise
+		 * return. */
+		if (new_threshold <= threshold)
+		{
+			elog(DEBUG2, "invalidation threshold for hypertable %d needs no update", hyper_id);
+			return false;
+		}
+
+		rel = table_open(threshold_relid, AccessExclusiveLock);
+
+		/* Fetch the latest threshold in case someone updated it
+		 * before we got exclusive lock. */
+		threshold = invalidation_threshold_fetch(rel, &tid);
+
+		if (new_threshold > threshold)
+		{
+			elog(DEBUG2,
+				 "updating existing invalidation threshold from " INT64_FORMAT " to " INT64_FORMAT,
+				 threshold,
+				 new_threshold);
+			invalidation_threshold_write(rel, hyper_id, new_threshold, &tid);
+			threshold_written = true;
+		}
+		else
+		{
+			elog(DEBUG2,
+				 "tried to set invalidation threshold to " INT64_FORMAT
+				 " but it was updated to " INT64_FORMAT " before getting update lock, skipping",
+				 new_threshold,
+				 threshold);
+		}
+	}
+	else
+	{
+		/* No threshold existed so we need to take an AccessExclusiveLock in
+		 * order to serialize all access to the threshold. */
+		elog(DEBUG2,
+			 "inserting new invalidation threshold " INT64_FORMAT " for hypertable %d",
+			 new_threshold,
+			 hyper_id);
+
+		rel = table_open(threshold_relid, AccessExclusiveLock);
+
+		/* Check if another transaction inserted a threshold before we got
+		 * exclusive lock on the table. */
+		threshold = invalidation_threshold_read(hyper_id, &tid, AccessExclusiveLock);
+
+		if (ItemPointerIsValid(&tid))
```

Comment:
FWIW, I have tests for these two cases in the concurrent refresh test. But they are currently commented because they are flaky since two concurrent refresh process race to get to this case, or the other one below. Still trying to figure out if there is a way to make them deterministic, but probably not possible.